File: backend/mr-db-sql/src/main/java/io/featurehub/db/services/PersonSqlApi.java
package io.featurehub.db.services;

import io.ebean.Database;
import io.ebean.FutureList;
import io.ebean.FutureRowCount;
import io.ebean.annotation.Transactional;
import io.featurehub.db.api.FillOpts;
import io.featurehub.db.api.NotNull;
import io.featurehub.db.api.OptimisticLockingException;
import io.featurehub.db.api.Opts;
import io.featurehub.db.api.PersonApi;
import io.featurehub.db.model.DbGroup;
import io.featurehub.db.model.DbOrganization;
import io.featurehub.db.model.DbPerson;
import io.featurehub.db.model.query.QDbPerson;
import io.featurehub.db.password.PasswordSalter;
import io.featurehub.mr.model.Group;
import io.featurehub.mr.model.Person;
import io.featurehub.mr.model.SortOrder;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import jakarta.inject.Inject;
import jakarta.inject.Singleton;
import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;
import java.util.UUID;
import java.util.concurrent.ExecutionException;
import java.util.stream.Collectors;

@Singleton
public class PersonSqlApi implements PersonApi {
  private static final Logger log = LoggerFactory.getLogger(PersonSqlApi.class);
  private final Database database;
  private final Conversions convertUtils;
  public final static int MAX_SEARCH = 100;
  private final ArchiveStrategy archiveStrategy;
  private final PasswordSalter passwordSalter = new PasswordSalter();

  @Inject
  public PersonSqlApi(Database database, Conversions convertUtils, ArchiveStrategy archiveStrategy) {
    this.database = database;
    this.convertUtils = convertUtils;
    this.archiveStrategy = archiveStrategy;
  }

  @Override
  public Person update(UUID id, Person person, Opts opts, UUID updatedBy) throws OptimisticLockingException {
    Conversions.nonNullPersonId(id);
    Conversions.nonNullPerson(person);
    Conversions.nonNullPersonId(updatedBy);

    DbPerson adminPerson = convertUtils.byPerson(updatedBy, Opts.opts(FillOpts.Groups));

    DbPerson p = convertUtils.byPerson(id, opts);

    if (adminPerson != null && p != null && p.getWhenArchived() == null) {
      return updatePerson(person, opts, adminPerson, p);
    }

    return null;
  }

  @Override
  public boolean noUsersExist() {
    return !new QDbPerson().exists();
  }

  public Person updatePerson(Person person, Opts opts, DbPerson adminPerson, DbPerson p) throws OptimisticLockingException {
    if (person.getVersion() == null || p.getVersion() != person.getVersion()) {
      throw new OptimisticLockingException();
    }

    if (person.getName() != null) {
      p.setName(person.getName());
    }
    if (person.getEmail() != null) {
      p.setEmail(person.getEmail());
    }

    if (person.getGroups() != null) {
      // we are going to need their groups to determine what they can do
      Person admin = convertUtils.toPerson(adminPerson, Opts.opts(FillOpts.Groups));

      boolean adminSuperuser = admin.getGroups().stream().anyMatch(g -> g.getPortfolioId() == null && g.getAdmin());
      List<UUID> validPortfolios = admin.getGroups()
        .stream()
        .filter(g -> g.getPortfolioId() != null && g.getAdmin())
        .map(Group::getPortfolioId)
        .collect(Collectors.toList());

      if (!adminSuperuser && validPortfolios.isEmpty()) {
        return null; // why are they even here???
      }

      // TODO: if groups are passed, limit removal of groups to the organisation relevant

      List<DbGroup> removeGroups = new ArrayList<>();
      List<UUID> replacementGroupIds = person.getGroups().stream().map(Group::getId).collect(Collectors.toList());
      List<UUID> foundReplacementGroupIds = new ArrayList<>(); // these are the ones we found

      p.getGroupsPersonIn().forEach(g -> {
        if (replacementGroupIds.contains(g.getId())) { // this has been passed to us and it is staying
          foundReplacementGroupIds.add(g.getId());
        } else if (adminSuperuser || (g.getOwningPortfolio() != null && validPortfolios.contains(g.getOwningPortfolio().getId()))) {
          // only if the admin is a superuser or this is a group in one of their portfolios will we honour it
          removeGroups.add(g);
        } else {
          log.warn("No permission to remove group {} from user {}", g, p.getEmail());
        }
      });

      log.debug("Removing groups {} from user {}", removeGroups, p.getEmail());
      // now remove them from these groups, we know this is valid
      p.getGroupsPersonIn().removeAll(removeGroups);

      replacementGroupIds.removeAll(foundReplacementGroupIds); // now this should be id's that we should consider adding

      log.debug("Attempting to add groups {} to user {} ", replacementGroupIds, p.getEmail());
      // now we have to find the replacement groups and see if this user is allowed to add them
      replacementGroupIds.forEach(gid -> {
        DbGroup group = convertUtils.byGroup(gid, Opts.empty());
        if (group != null && (adminSuperuser ||
          (group.getOwningPortfolio() != null && validPortfolios.contains(group.getOwningPortfolio().getId())) )) {
          p.getGroupsPersonIn().add(group);
        } else {
          log.warn("No permission to add group {} to user {}", group, p.getEmail());
        }
      });
    }

    updatePerson(p);

    return convertUtils.toPerson(p, opts);
  }

  @Override
  public PersonPagination search(String filter, SortOrder sortOrder, int offset, int max, Opts opts) {
    offset = Math.max(offset, 0);

    max = Math.min(max, MAX_SEARCH);
    max = Math.max(max, 1);

    // set the limits
    QDbPerson search = new QDbPerson().setFirstRow(offset).setMaxRows(max);

    // set the filter if anything, make sure it is case insignificant
    if (filter != null) {
      // name is mixed case, email is always lower case
      search = search.or().name.icontains(filter).email.contains(filter.toLowerCase()).endOr();
    }


    if (sortOrder != null) {
      if (sortOrder == SortOrder.ASC) {
        search = search.order().name.asc();
      } else {
        search = search.order().name.desc();
      }
    }

    if (!opts.contains(FillOpts.Archived)) {
      search = search.whenArchived.isNull();
    }

    FutureRowCount<DbPerson> futureCount = search.findFutureCount();
    FutureList<DbPerson> futureList = search.findFutureList();

    PersonPagination pagination = new PersonPagination();

    try {
      pagination.max = futureCount.get();
      DbOrganization org = convertUtils.getDbOrganization();
      final List<DbPerson> dbPeople = futureList.get();
      pagination.people = dbPeople.stream().map(dbp ->
        convertUtils.toPerson(dbp, org, opts)
      ).collect(Collectors.toList());

      LocalDateTime now = LocalDateTime.now();

      pagination.personIdsWithExpiredTokens = dbPeople.stream()
        .filter(p -> p.getToken() != null && p.getTokenExpiry() != null && p.getTokenExpiry().isBefore(now))
        .map(DbPerson::getId)
        .collect(Collectors.toList());

      pagination.personsWithOutstandingTokens = dbPeople.stream()
        .filter(p -> p.getToken() != null)
        .map(p -> new PersonToken(p.getToken(), p.getId()))
        .collect(Collectors.toList());

      return pagination;
    } catch (InterruptedException | ExecutionException e) {
      log.error("Failed to execute search.", e);
      return null;
    }
  }

  @Override
  public Person get(@NotNull String email, Opts opts) {
    if (email == null) {
      throw new IllegalArgumentException("email required");
    }

    if (email.contains("@")) {
      QDbPerson search = new QDbPerson().email.eq(email.toLowerCase());
      if (!opts.contains(FillOpts.Archived)) {
        search = search.whenArchived.isNull();
      }
      return search.groupsPersonIn.fetch()
        .findOneOrEmpty()
        .map(p -> convertUtils.toPerson(p, opts))
        .orElse(null);
    }

    UUID id = Conversions.checkUuid(email);

    if (id != null) {
      return get(id, opts);
    }

    return null;
  }

  @Override
  public Person get(UUID id, Opts opts) {
    Conversions.nonNullPersonId(id);


    QDbPerson search = new QDbPerson().id.eq(id);
    if (!opts.contains(FillOpts.Archived)) {
      search = search.whenArchived.isNull();
    }
    return search.groupsPersonIn.fetch()
      .findOneOrEmpty()
      .map(p -> convertUtils.toPerson(p, opts))
      .orElse(null);

  }

  @Override
  public Person getByToken(String token, Opts opts) {
    if (token == null) return null;

    DbPerson person  = new QDbPerson().whenArchived.isNull().token.eq(token).findOne();

    if (person != null && person.getTokenExpiry().isAfter(getNow())) {
      return convertUtils.toPerson(person, opts);
    }
    return null;
  }

  protected LocalDateTime getNow() {
    return LocalDateTime.now();
  }

  @Override
  public PersonToken create(String email, String name, UUID createdBy) throws DuplicatePersonException {
    if (email == null || name == null) {
      return null;
    }

    DbPerson created = createdBy == null ? null : convertUtils.byPerson(createdBy);

    if (createdBy != null && created == null) {
      return null;
    }

    PersonToken personToken;
    final DbPerson onePerson = new QDbPerson().email.eq(email.toLowerCase()).findOne();
    if (onePerson == null) {

      String token = UUID.randomUUID().toString();
      DbPerson.Builder builder = new DbPerson.Builder()
        .email(email.toLowerCase())
        .name(name)
        .token(token)
        .tokenExpiry(getNow().plusDays(7));

      if (created != null) {
        builder.whoCreated(created);
      }

      DbPerson person = builder.build();
      updatePerson(person);

      personToken = new PersonToken(person.getToken(), person.getId());
    } else if (onePerson.getWhenArchived() != null) {
      onePerson.setWhenArchived(null);
      onePerson.setToken(UUID.randomUUID().toString()); // ensures it gets past registration again
      onePerson.setName(name);
      if (created != null) {
        onePerson.setWhoCreated(created);
      }
      updatePerson(onePerson);
      return null;
    } else {
      throw new DuplicatePersonException();
    }

    return personToken;
  }

  /**
   * This person will be fully formed, not token. Usually used only for testing.
   */
  Person createPerson(String email, String name, String password, UUID createdById, Opts opts) throws DuplicatePersonException {
    if (email == null) {
      return null;
    }

    DbPerson created = createdById == null ? null : convertUtils.byPerson(createdById);

    if (createdById != null && created == null) {
      return null;
    }

    if (new QDbPerson().email.eq(email.toLowerCase()).findOne() == null) {

      DbPerson.Builder builder = new DbPerson.Builder()
        .email(email.toLowerCase())
        .name(name);

      if (created != null) {
        builder.whoCreated(created);
      }

      DbPerson person = builder.build();
      passwordSalter.saltPassword(password).ifPresent(person::setPassword);

      updatePerson(person);

      return convertUtils.toPerson(person, opts);
    } else {
      throw new DuplicatePersonException();
    }
  }


  @Transactional
  private void updatePerson(DbPerson p) {
    database.save(p);
  }

  @Override
  public boolean delete(String email) {
    return new QDbPerson().email.eq(email.toLowerCase()).findOneOrEmpty().map(p -> {
      archiveStrategy.archivePerson(p);
      return true;
    }).orElse(false);
  }
}

File: backend/mr-db-sql/src/main/kotlin/io/featurehub/db/services/PersonSqlApi.kt
package io.featurehub.db.services

import io.ebean.Database
import io.ebean.annotation.Transactional
import io.ebean.annotation.TxType
import io.featurehub.db.api.*
import io.featurehub.db.model.DbGroupMember
import io.featurehub.db.model.DbGroupMemberKey
import io.featurehub.db.model.DbLogin
import io.featurehub.db.model.DbPerson
import io.featurehub.db.model.query.*
import io.featurehub.db.password.PasswordSalter
import io.featurehub.mr.model.*
import jakarta.inject.Inject
import jakarta.inject.Singleton
import org.apache.commons.lang3.RandomStringUtils
import org.slf4j.LoggerFactory
import java.time.Instant
import java.time.LocalDateTime
import java.util.*
import java.util.concurrent.ExecutionException

@Singleton
open class PersonSqlApi @Inject constructor(
  private val database: Database,
  private val convertUtils: Conversions,
  private val archiveStrategy: ArchiveStrategy,
  private val internalGroupSqlApi: InternalGroupSqlApi
) : PersonApi, InternalPersonApi {
  private val passwordSalter = PasswordSalter()

  @Throws(OptimisticLockingException::class)
  override fun update(id: UUID, person: Person, opts: Opts, updatedBy: UUID): Person? {
    val version = person.version ?: return null

    val dbPerson = updatePersonDetails(id, updatedBy, version, person.name,
      person.email, person.groups?.mapNotNull { it.id }, false)

    return if (dbPerson != null) convertUtils.toPerson(dbPerson, opts) else null
  }

  override fun noUsersExist(): Boolean {
    return !QDbPerson().exists()
  }


  inner class GroupChangeCollection {
    val groupsToAdd = mutableListOf<UUID>()
    val groupsToRemove = mutableListOf<UUID>()

    override fun toString(): String {
      return "to remove $groupsToRemove - to add $groupsToAdd"
    }
  }


  fun updatePersonDetails(personId: UUID, updatedBy: UUID, version: Long, name: String?, email: String?, groups: List<UUID>?, unarchive: Boolean) : DbPerson? {
    val updatingRecord = QDbPerson().id.eq(personId).findOne() ?: return null

    if (updatingRecord.personType == PersonType.SDKSERVICEACCOUNT) {
      return null
    }

    if (version != updatingRecord.version) {
      throw OptimisticLockingException()
    }

    name?.let {
      updatingRecord.name = it
    }

    if (updatingRecord.personType != PersonType.SERVICEACCOUNT) {
      val organization = convertUtils.dbOrganization()
      val groupChanges = GroupChangeCollection()
      val superuserChanges = SuperuserChanges(organization)

      // reactivate user?
      if (unarchive && updatingRecord.whenArchived != null) {
        updatingRecord.whenArchived = null
      }

      email?.let {
        updatingRecord.email = it
      }

      groups?.let { groupsTheyWant ->
        val adminGroupsUpdatingPersonBelongsTo = internalGroupSqlApi.adminGroupsPersonBelongsTo(updatedBy)

        val performedBySuperuser = adminGroupsUpdatingPersonBelongsTo.any { it.owningPortfolio == null }

        val portfoliosPerformingUserCanManage = adminGroupsUpdatingPersonBelongsTo
          .mapNotNull { it.owningPortfolio?.id }
          .toList()

        if (!performedBySuperuser && portfoliosPerformingUserCanManage.isEmpty()) {
          return null // why are they even here???
        }

        val superuserGroup = internalGroupSqlApi.superuserGroup(organization)

        val groupsAlreadyIn = QDbGroupMember()
          .select(QDbGroupMember.Alias.id.groupId)
          .person.id.eq(personId)
          .findList().map { it.id.groupId!! }


        // we start with a list of groups they want and remove the list of groups they are already in
        // and that gives us the list of groups they want to add
        val groupsTheyWantToAdd = with(groupsTheyWant.toMutableList()) {
          removeAll(groupsAlreadyIn)
          this
        }

        if (!performedBySuperuser) { // if the superuser isn't doing this, we need to check if its ok
          removeChangeIfOutsidePortfolioPermission(groupsTheyWantToAdd, portfoliosPerformingUserCanManage)
        }

        if (groupsTheyWantToAdd.contains(superuserGroup?.id)) {
          superuserChanges.addedSuperuserPersonIds.add(personId)
        }

        // we start with a list of all the groups they are already in and subtract the groups they have told us they want.
        // the difference is the groups they no longer want
        val groupsTheyWantToRemove = with(groupsAlreadyIn.toMutableList()) {
          removeAll(groupsTheyWant)
          this
        }

        if (!performedBySuperuser) {
          removeChangeIfOutsidePortfolioPermission(groupsTheyWantToRemove, portfoliosPerformingUserCanManage)
        }

        if (groupsTheyWantToRemove.contains(superuserGroup?.id)) {
          superuserChanges.removedSuperusers.add(personId)
        }

        groupChanges.groupsToAdd.addAll(groupsTheyWantToAdd)
        groupChanges.groupsToRemove.addAll(groupsTheyWantToRemove)

        log.debug("Changing groups for person ${personId} as $groupChanges")
      }

      updatePerson(updatingRecord, groupChanges, superuserChanges)
    } else {
      updatePerson(updatingRecord, null, null)
    }

    return updatingRecord
  }

  override fun updateV2(personId: UUID, updatePerson: UpdatePerson, updatedBy: UUID): UUID? {
    val person = updatePersonDetails(personId, updatedBy, updatePerson.version, updatePerson.name,
      updatePerson.email, updatePerson.groups, updatePerson.unarchive == true)

    return if (person == null) null else personId
  }

  private fun removeChangeIfOutsidePortfolioPermission(
    groupsTheyWantToAdd: MutableList<UUID>,
    portfoliosPerformingUserCanManage: List<UUID>
  ) {
    val groupPortfolioMap =
      QDbGroup().id.`in`(groupsTheyWantToAdd)
        .select(QDbGroup.Alias.id, QDbGroup.Alias.owningPortfolio.id).findList()
        .associate { it.id!! to it.owningPortfolio.id }

    // now we have to work through them to ensure they are allowed to add them
    groupsTheyWantToAdd.removeIf { groupId ->
      !portfoliosPerformingUserCanManage.contains(groupPortfolioMap[groupId])
    }
  }

  val nilUUID = UUID.fromString("00000000-0000-0000-0000-000000000000")

  override fun search(
    filter: String?, sortOrder: SortOrder?, offset: Int, max: Int,
    pTypes: Set<PersonType?>,
    sortBy: SearchPersonSortBy?,
    opts: Opts
  ): PersonApi.PersonPagination {
    val searchOffset = offset.coerceAtLeast(0)
    val searchMax = max.coerceAtMost(MAX_SEARCH).coerceAtLeast(1)

    // set the limits
    var search = QDbPerson()
      .id.ne(nilUUID)

    // set the filter if anything, make sure it is case insignificant
    if (filter != null) {
      val fil = filter.lowercase(Locale.getDefault())
      // name is mixed case, email is always lower case
      search = search.or().name.icontains(fil).email.contains(fil).endOr()
    }

    val personTypes = pTypes.filter { it != null && it != PersonType.SDKSERVICEACCOUNT }

    search = if (personTypes.isNotEmpty()) {
      if (personTypes.size == 1) search.personType.eq(personTypes.first()) else search.personType.`in`(personTypes)
    } else {
      search.personType.eq(PersonType.PERSON)
    }

    if (sortOrder != null) {
      if (sortBy == null || sortBy == SearchPersonSortBy.NAME) {
        search = if (sortOrder == SortOrder.ASC) {
          search.orderBy("upper(name) asc")
        } else {
          search.orderBy("upper(name) desc")
        }
      } else if (sortBy == SearchPersonSortBy.ACTIVATIONSTATUS) {
        search = if (sortOrder == SortOrder.ASC) {
          search.orderBy("(case when when_archived is null then 0 else 1 end) asc, upper(name) asc")
        } else {
          search.orderBy("(case when when_archived is null then 0 else 1 end) desc, upper(name) desc")
        }
      }
    }

    if (!opts.contains(FillOpts.Archived)) {
      search = search.whenArchived.isNull
    }

    val futureCount = search.findFutureCount()
    val futureList = search
      .setFirstRow(searchOffset).setMaxRows(searchMax)
      .findFutureList()

    return try {
      val org = convertUtils.dbOrganization()
      val dbPeople = futureList.get()
      val people = dbPeople.map { dbp: DbPerson? -> convertUtils.toPerson(dbp, org, opts)!! }

      val peopleGroupCount: Map<UUID, Int> =
        if (opts.contains(FillOpts.CountGroups)) findGroupCounts(dbPeople) else emptyMap()

      PersonApi.PersonPagination(
        futureCount.get(),
        if (opts.contains(FillOpts.CountGroups)) people.map { p -> toSearchPerson(p, peopleGroupCount) } else emptyList(),
        if (!opts.contains(FillOpts.CountGroups)) people else emptyList(),
        dbPeople
          .filter { p -> p.token != null }
          .map { p -> PersonApi.PersonToken(p.token, p.id) },
        dbPeople
          .filter { p -> p.token != null && p.tokenExpiry != null && p.tokenExpiry.isBefore(now) }
          .map { obj -> obj.id },
      )
    } catch (e: InterruptedException) {
      log.error("Failed to execute search.", e)
      throw RuntimeException(e)
    } catch (e: ExecutionException) {
      log.error("Failed to execute search.", e)
      throw RuntimeException(e)
    }
  }

  private fun findGroupCounts(dbPeople: List<DbPerson>) =
    QGroupMemberAgg()
      .select(QGroupMemberAgg.Alias.personId, QGroupMemberAgg.Alias.counter)
      .personId.`in`(dbPeople.map { it.id }).findList().associate {
        it.personId to it.counter
      }

  override fun get(email: String, opts: Opts): Person? {
    if (email.contains("@")) {
      var search = QDbPerson().email.eq(email.lowercase(Locale.getDefault()))
      if (!opts.contains(FillOpts.Archived)) {
        search = search.whenArchived.isNull
      }
      return search.groupMembers.fetch()
        .findOneOrEmpty()
        .map { p: DbPerson? -> convertUtils.toPerson(p, opts) }
        .orElse(null)
    }
    val id = Conversions.checkUuid(email)
    return id?.let { get(it, opts) }
  }

  override fun get(id: UUID, opts: Opts): Person? {
    var search = QDbPerson().id.eq(id)

    if (!opts.contains(FillOpts.Archived)) {
      search = search.whenArchived.isNull
    }

    return search.groupMembers.fetch()
      .findOne()?.let { convertUtils.toPerson(it, opts) }
  }

  override fun getByToken(token: String, opts: Opts): Person? {
    val person = QDbPerson().whenArchived.isNull.token.eq(token).findOne()
    return if (person != null && person.tokenExpiry.isAfter(now)) {
      convertUtils.toPerson(person, opts)!!
    } else null
  }

  protected val now: LocalDateTime get() = LocalDateTime.now()

  @Throws(PersonApi.DuplicatePersonException::class)
  override fun create(email: String, name: String?, createdBy: UUID?): PersonApi.PersonToken? {
    val created = if (createdBy == null) null else convertUtils.byPerson(createdBy)
    if (createdBy != null && created == null) {
      return null
    }
    val personToken: PersonApi.PersonToken
    val onePerson = QDbPerson().email.eq(email.lowercase(Locale.getDefault())).findOne()
    if (onePerson == null) {
      val token = UUID.randomUUID().toString()
      val builder = DbPerson.Builder()
        .email(email.lowercase(Locale.getDefault()))
        .name(name)
        .token(token)
        .tokenExpiry(now.plusDays(7))
      if (created != null) {
        builder.whoCreated(created)
      }
      val person = builder.build()
      updatePerson(person)
      personToken = PersonApi.PersonToken(person.token, person.id)
    } else if (onePerson.whenArchived != null) {
      onePerson.whenArchived = null
      onePerson.token = UUID.randomUUID().toString() // ensures it gets past registration again
      onePerson.name = name
      onePerson.isPasswordRequiresReset = true
      if (created != null) {
        onePerson.whoCreated = created
      }
      updatePerson(onePerson)
      personToken = PersonApi.PersonToken(onePerson.token, onePerson.id)
    } else {
      throw PersonApi.DuplicatePersonException()
    }
    return personToken
  }

  @Transactional(type = TxType.REQUIRES_NEW)
  override fun createServicePerson(name: String, createdBy: UUID?): CreatedServicePerson? {
    val created = if (createdBy == null) null else convertUtils.byPerson(createdBy)
    if (createdBy != null && created == null) {
      return null
    }

    val servicePerson = DbPerson()
    servicePerson.name = name
    with(servicePerson) {
      id = UUID.randomUUID()
      email = "${id}@admin.sa.featurehub.io"
      personType = PersonType.SERVICEACCOUNT
      whoChanged = created
      save()
    }

    val token = createAdminServiceAccountToken(servicePerson)

    return CreatedServicePerson(convertUtils.toPerson(servicePerson)!!, token)
  }

  private fun createAdminServiceAccountToken(person: DbPerson): String {
    val token = RandomStringUtils.randomAlphanumeric(48)
    DbLogin.Builder().person(person).token(token).lastSeen(Instant.now()).build().save()
    return token
  }

  @Transactional(type = TxType.REQUIRES_NEW)
  override fun resetServicePersonToken(serviceAccountId: UUID): CreatedServicePerson? {
    val search = QDbPerson().id.eq(serviceAccountId)
      .personType.eq(PersonType.SERVICEACCOUNT)
      .whenArchived.isNull.findOne()

    if (search != null) {
      QDbLogin().person.eq(search).delete()
      val token = createAdminServiceAccountToken(search)
      return CreatedServicePerson(convertUtils.toPerson(search)!!, token)
    }

    return null
  }

  /**
   * This person will be fully formed, not token. Usually used only for testing.
   */
  @Throws(PersonApi.DuplicatePersonException::class)
  fun createPerson(email: String?, name: String?, password: String?, createdById: UUID?, opts: Opts?): Person? {
    if (email == null) {
      return null
    }
    val created = if (createdById == null) null else convertUtils.byPerson(createdById)
    if (createdById != null && created == null) {
      return null
    }
    return if (!QDbPerson().email.eq(email.lowercase(Locale.getDefault())).exists()) {
      val builder = DbPerson.Builder()
        .email(email.lowercase(Locale.getDefault()))
        .name(name)
      if (created != null) {
        builder.whoCreated(created)
      }
      val person = builder.build()
      passwordSalter.saltPassword(password, DbPerson.DEFAULT_PASSWORD_ALGORITHM)
        .ifPresent { pwd: String? -> person.password = pwd }
      person.passwordAlgorithm = DbPerson.DEFAULT_PASSWORD_ALGORITHM
      updatePerson(person)
      convertUtils.toPerson(person, opts!!)
    } else {
      throw PersonApi.DuplicatePersonException()
    }
  }

  @Transactional(type = TxType.REQUIRES_NEW)
  private fun updatePerson(
    p: DbPerson,
    groupChangeCollection: GroupChangeCollection? = null,
    superuserChanges: SuperuserChanges? = null
  ) {
    database.save(p)

    if (groupChangeCollection != null) {
      if (groupChangeCollection.groupsToRemove.isNotEmpty()) {
        QDbGroupMember().person.id.eq(p.id).group.id.`in`(groupChangeCollection.groupsToRemove).delete()
      }

      if (groupChangeCollection.groupsToAdd.isNotEmpty()) {
        groupChangeCollection.groupsToAdd.forEach { g ->
          if (!QDbGroupMember().person.id.eq(p.id).group.id.eq(g).exists()) {
            DbGroupMember(DbGroupMemberKey(p.id, g)).save()
          }
        }
      }
    }

    if (superuserChanges != null) {
      internalGroupSqlApi.updateSuperusersFromPortfolioGroups(superuserChanges)
    }
  }

  override fun delete(email: String, deleteGroups: Boolean): Boolean {
    return QDbPerson().email.eq(email.lowercase(Locale.getDefault())).findOne()?.let { p ->
      if (p.personType == PersonType.SDKSERVICEACCOUNT) {
        false  // can't delete sdk service accounts even if you know their email!
      } else {
        archiveStrategy.archivePerson(p)
        // remove all of their tokens
        QDbLogin().person.id.eq(p.id).delete()
        if (deleteGroups) {
          QDbGroupMember().person.id.eq(p.id).delete()
        }
        true
      }
    } ?: false
  }


  companion object {
    private val log = LoggerFactory.getLogger(PersonSqlApi::class.java)
    const val MAX_SEARCH = 100

    fun toSearchPerson(person: Person, groupCountsByPersonId: Map<UUID, Int>): SearchPerson {
      return SearchPerson()
        .personType(person.personType!!)
        .id(person.id!!.id)
        .email(person.email!!)
        .name(person.name!!)
        .version(person.version!!)
        .whenLastAuthenticated(person.whenLastAuthenticated)
        .whenDeactivated(person.whenArchived)
        .whenLastSeen(person.whenLastSeen)
        .groupCount(groupCountsByPersonId[person.id!!.id] ?: 0)
    }
  }

  @Transactional(type = TxType.REQUIRES_NEW)
  override fun createSdkServiceAccountUser(name: String, createdBy: DbPerson, archived: Boolean): DbPerson {
    val servicePerson = DbPerson()
    servicePerson.name = name
    with(servicePerson) {
      id = UUID.randomUUID()
      email = "${id}@sdk.sa.featurehub.io"
      personType = PersonType.SDKSERVICEACCOUNT
      whoChanged = createdBy
      if (archived) { // only used when migrating old service accounts
        whenArchived = now
      }
      save()
    }

    return servicePerson
  }


  override fun deleteSdkServiceAccountUser(personId: UUID, createdBy: DbPerson) {
    QDbPerson().id.eq(personId).findOne()?.let {
      it.whoChanged = createdBy
      it.whenArchived = now
      it.save()
    }
  }

  override fun updateSdkServiceAccountUser(personId: UUID, updatedBy: DbPerson, name: String) {
    QDbPerson().id.eq(personId)
      .asUpdate()
      .set(QDbPerson.Alias.name, name)
      .set(QDbPerson.Alias.whoChanged, updatedBy)
      .update()
  }

  override fun findSuperUserToBlame(orgId: UUID): DbPerson {
    val superuserGroup = internalGroupSqlApi.superuserGroup(orgId)!!

    return QDbPerson().select(QDbPerson.Alias.id).groupMembers.group.id.eq(superuserGroup.id).setMaxRows(1).findList().first()
  }
}
